### ISSUE TITLE
[DataGrid] Export - Add custom format

### ISSUE THREAD

**Issue by jy95** [mui-org/material-ui-x#1308]
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Summary 💡

To be able to add some code so that we could export data into another format (not) yet defined such as ICAL

## Examples 🌈

UI side, It would mean that we add user defined export possibilities :  

![feature](https://user-images.githubusercontent.com/9306961/112748782-75eda980-8fbe-11eb-9ccd-81598882f070.PNG)

Code side, it would mean : 
- We control the [MIME type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types) of the generated file
- We control the name of the generated file
- We provide the content of the generated file ([ICalendar](https://icalendar.org/) is a structured format) based on the data we have
- We provide a label of this option (and it will vary depending of the language selected)

## Motivation 🔦

I have a Youtube channel and it would be nice that I could give the possibility to export my schedule into a ical file (so that it could be previewed into any calendar app).
Currently, I displayed my schedule here : https://jy95.github.io/yt_gaming_library/planning

**Comment by oliviertassinari** [comment]
I wonder if this is not an opportunity to expose a new API method, to breakdown

https://github.com/mui-org/material-ui-x/blob/82f3d8485852b2fabfa9592b25d850d966fb8705/packages/grid/_modules_/grid/hooks/features/export/useGridCsvExport.tsx#L22

further. Would it make sense to have an export feature? Something that returns JavaScript input used for the CSV export. It would apply the filters, sorting, pagination, column selection transformations.

**Comment by jy95** [comment]
Maybe (as apiRef is your internal api) : I though about something that is complementary :

```js
import { DataGrid, GridToolbar } from "@material-ui/data-grid";

// common attributes for data grid
const data = [
  /* ... */
];
const columns = [
  /* ... */
];

// for users who wants translations for their custom export option
const translator = (_) => "Export as ICAL";
const customLocaleText = (language) => ({ exportAsICal: translator(language) });

// So initilization could be :
return (
  <DataGrid
    rows={data}
    columns={columns}
    disableSelectionOnClick
    //disableExtendRowFullWidth // No needed for now
    disableColumnFilter // or filterable: false in each column
    autoHeight
    localeText={customLocaleText}
    components={{
      Toolbar: GridToolbar,
      customExports: [
        {
          translationKey: "exportAsICal",
          // some function that returns the content of the file
          createContent: (visibleColumns, visibleSortedRows) => {
            // TODO
            return "";
          },
          filename: "games-schedule.ical",
          type: "text/calendar",
        },
      ],
    }}
  />
);
```

With that way, it will be easy to reuse your `exportAs `function, no ?

**Comment by oliviertassinari** [comment]
> as apiRef is your internal API

It's public in XGrid.

> localeText={customLocaleText}

This is not extendable. You would need to provide your own translation

> customExports

I think that the pattern could work for developers that build their own toolbar, as a prop of `GridToolbarExport`.

**Comment by flaviendelangle** [comment]
@alexfauquette I think #3891 closes this one

**Comment by alexfauquette** [comment]
Yes, in next release, it will be possible to add your own options to the export menu as follow:

```jsx
const MyToolbarExport = ({ csvOptions, printOptions, ...other }) => (
  <GridToolbarExportContainer {...other}>
    <GridCsvExportMenuItem options={csvOptions} />
    <GridPrintExportMenuItem options={printOptions} />
    <MyIclaExportMenuItem />
  </GridToolbarExportContainer>
);
```

With `<MyIclaExportMenuItem />` receiving a prop `hideMenu` to close the export menu after the export as follow

```jsx
const MyIclaExportMenuItem = ({hideMenu}) => <MenuItem
		onClick={() => {
			// get and export the data as you want
			hideMenu()
		}}
	>
		ICLA
	</MenuItem>
```

More detail will be available in the doc [here is the perview](https://deploy-preview-3891--material-ui-x.netlify.app/components/data-grid/export/#custom-export-format)
